handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -917,16 +927,19 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
         discardSomeReadBytes();
 
         flushIfNeeded(ctx);
+        readIfNeeded(ctx);
+
+        firedChannelRead = false;
+        ctx.fireChannelReadComplete();
+    }
 
+    private void readIfNeeded(ChannelHandlerContext ctx) {
         // If handshake is not finished yet, we need more data.
         if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
             // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
             // yet, which means we need to trigger the read to ensure we not encounter any stalls.
             ctx.read();
```

Comment:
We can not remove throws Exception here as this would break a users handler that extend SslHandler and override read(...) and still have the throws in the signature. But anyway let us leave it as it is for now and investigate later

> Am 28.03.2016 um 22:40 schrieb Scott Mitchell <notifications@github.com>:
> 
> In handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
> 
> >  
> > +    private void readIfNeeded(ChannelHandlerContext ctx) {
> >          // If handshake is not finished yet, we need more data.
> >          if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
> >              // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
> >              // yet, which means we need to trigger the read to ensure we not encounter any stalls.
> >              ctx.read();
> I'd probably need to add throws Exception for every single method in the call-graph
> 
> @vkostyukov - You should be able to remove throws Exception from the signature of public void read(ChannelHandlerContext ctx) ... however to your point it seems reasonable that a read "synthesized" by SslHandler should not force a read after the handshake is complete.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>